vendor/engines/secure_rooms/app/controllers/secure_rooms_api/scans_controller.rb

```diff
@@ -5,51 +5,40 @@ class SecureRoomsApi::ScansController < ApplicationController
     password: Settings.secure_rooms_api.basic_auth_password,
   )
 
-  rescue_from ActiveRecord::RecordNotFound, with: :report_missing_ids
-
-  before_action :load_models
+  before_action :load_user_and_reader
 
   def scan
-    user_accounts = @user.accounts_for_product(@card_reader.secure_room)
-
-    if user_accounts.present?
-      response_status = user_accounts.many? ? :multiple_choices : :ok
-
-      # TODO: needs a legitimage tablet_identifier once tablet exists
-      response_json = {
-        response: "select_account",
-        tablet_identifier: "abc123",
-        name: @user.full_name,
-        accounts: SecureRooms::AccountPresenter.wrap(user_accounts),
-      }
-    else
-      response_status = :forbidden
-      response_json = {
-        response: "deny",
-        reason: "No accounts found",
-      }
-    end
-
-    render json: response_json, status: response_status
-  end
-
-  def load_models
-    @user = User.find_by!(card_number: params[:card_number])
-    @card_reader = SecureRooms::CardReader.find_by!(
-      card_reader_number: params[:reader_identifier],
-      control_device_number: params[:controller_identifier],
+    accounts = @user.accounts_for_product(@card_reader.secure_room)
+    selected_account = accounts.find { |account| account.id == params[:account_identifier] }
+
+    access_verdict = SecureRooms::CheckAccess.new.authorize(
+      @user,
+      @card_reader,
+      accounts,
+      selected_account,
     )
+
+    render json: build_json(access_verdict, accounts), status: access_verdict.http_status
   end
 
   private
 
-  def report_missing_ids(error)
-    response_json = {
-      response: "deny",
-      reason: error.message,
+  def build_json(access_verdict, accounts)
+    {
+      # TODO: (#140895375) return actual tablet_identifier
+      tablet_identifier: "abc123",
+      name: @user.full_name,
+      reason: access_verdict.reason,
```

Comment:
Should we keep the `response` field as well so you can see it in the JSON as well as the HTTP status?